### PR TITLE
add a time-to-RFC3339 conversion function

### PR DESCRIPTION
**Pull request by c-cube**

### CODE REVIEW

File: core/time/rfc3339.odin [hunk 1]
```diff
@@ -187,4 +187,108 @@ scan_digits :: proc(s: string, sep: string, count: int) -> (res: int, ok: bool)
...
+		}
+		offset += uint(n_digits)
+	}
+
+	if utc_offset == 0 {
+		temp_string[offset] = 'Z'
+		offset += 1
+	} else {
+		temp_string[offset] = utc_offset > 0 ? '+' : '-'
+		offset += 1;
```
Comment (flysand7) [inline core/time/rfc3339.odin:278]:
```suggestion
		offset += 1
```

File: tests/core/time/test_core_time.odin [hunk 2]
```diff
@@ -196,6 +196,35 @@ test_parse_rfc3339_string :: proc(t: ^testing.T) {
 	}
 }
 
+@test
+test_print_rfc3339 :: proc(t: ^testing.T) {
+	TestCase :: struct {
+		printed: string,
+		time: i64,
+		utc_offset: int
```
Comment (flysand7) [inline tests/core/time/test_core_time.odin:204]:
```suggestion
		utc_offset: int,
```

File: tests/core/time/test_core_time.odin [hunk 3]
```diff
@@ -196,6 +196,35 @@ test_parse_rfc3339_string :: proc(t: ^testing.T) {
...
+	for test in tests {
+		timestamp := time.Time { _nsec = test.time }
+		printed_timestamp, ok := time.time_to_rfc3339(time=timestamp, utc_offset=test.utc_offset)
+		defer delete_string(printed_timestamp)
+
+		testing.expect(t, ok, "expected printing to work fine")
+
+		testing.expectf(
+			t, printed_timestamp == test.printed,
+			"expected is %w, printed is %w", test.printed, printed_timestamp)
```
Comment (flysand7) [inline tests/core/time/test_core_time.odin:224]:
```suggestion
			"expected is %w, printed is %w", test.printed, printed_timestamp,
		)
```

File: core/time/rfc3339.odin [hunk 4]
```diff
@@ -187,4 +187,108 @@ scan_digits :: proc(s: string, sep: string, count: int) -> (res: int, ok: bool)
...
+- `utc_offset`: offset in minutes wrt UTC (ie. the timezone)
+- `include_nanos`: whether to include nanoseconds in the result.
+*/
+time_to_rfc3339 :: proc(time: Time, utc_offset : int = 0, include_nanos := true, allocator := context.allocator) -> (res: string, ok: bool) {
+	// convert to datetime
+	datetime := time_to_datetime(time) or_return
+
+	if datetime.year < 0 || datetime.year >= 10_000 { return "", false }
+
+	temp_string := [30]u8{}
```
Comment (flysand7) [inline core/time/rfc3339.odin:207]:
```suggestion
	temp_string := [36]u8{}
```

Date: 10 characters
`T`: 1 character
Time: 8 characters
`.`: 1 character
Nanos: 9 characters max
`Z`: 1 character
tz offset: 6 characters

36 characters maximum for an RFC3339 timestamp.

File: core/time/rfc3339.odin [hunk 5]
```diff
@@ -187,4 +187,108 @@ scan_digits :: proc(s: string, sep: string, count: int) -> (res: int, ok: bool)
...
+	offset : uint = 0 // offset in temp_string
```
Comment (flysand7) [inline core/time/rfc3339.odin:208]:
```suggestion
	offset : uint = 0
```

This comment's kind rudimentary

**Comment by Kelimion** [comment]
We run our tests with `-vet -strict-style`. The CI is complaining that you've shadowed the offset parameter.
```odin
D:/a/Odin/Odin/core/time/rfc3339.odin(279:3) Error: Declaration of 'utc_offset' shadows declaration at line 201 
	utc_offset := abs(utc_offset) 
	^ 
```
I would rename the `offset` you're using to keep track of where you are in the buffer to `index` or `idx`. Then you can write `offset := abs(utc_offset)` and use that for the TZ.

Alternatively you could write `utc_offset := utc_offset` at the top of the procedure to let the compiler know you're intentionally shadowing what is otherwise an immutable parameter.

**Comment by c-cube** [comment]
Thank you, I took the second option.

**Pull request merged**